[WIP] Supporting Elekta log files

SimonBiggs opened the pull request:
Hi @jrkerns,

We have successfully set up high resolution logfile exporting and decoding from an Elekta linac. I have created a module called [trf2csv](https://gitlab.com/level-0/trf2csv)  which converts the Elekta `.trf` files to a basic `.csv` file.

I am proposing to write a pull request to pylinac to support Elekta logfile analysis within your `Log Analyzer` module. I am aiming to do this piece by piece in collaboration with you so that I align with the practices you have employed within pylinac.

Is this okay with you?

SimonBiggs commented:
Ping @msobolewski

SimonBiggs commented:
So maybe I jumped the gun a bit. I think I might develop a few little modules achieving what I am after, then towards the end we can pull those modules in if need be.

jrkerns commented:
This project sounds awesome. Assuming the generation of logs is easy enough for most clinics to do, I'd love to look into this. I am worried however about the AGPLv3 license.

SimonBiggs commented:
I've detailed the log file extraction at the end of the readme:
https://gitlab.com/level-0/trf2csv#setup-logfile-extraction

It is a little arduous.

With regards to the AGPL I do have a few points. In Australia people's ability to share code is limited. In the public system the answer is very much you cannot share code unless there is an already existing license requiring you to do so. In the private system one of the worries is that if code is released then someone will take it, run with it, and make a product to compete with something you could have created.

I've thought long and hard and I have settled on the AGPL because this means someone can't just wrap up the code in a web interface without sharing their code. And anyone who really wants to give back but just have their hands tied now can actually do so because the license dictates that all code distributed is under the same license.

Orthanc, a dicom web server, has some parts under AGPL and some parts not. It achieves this via a plugin system. Instead of making trf2csv a requirement, could it be designed to be a sort of plugin? 

Also, so that you know where we are coming from, our goal is to do a comparison between Mosaiq SQL entries and the logfiles. And I want to do this over every delivery. There's quite a few steps required to achieve this. Each step I'm spinning out into its own module so that it can be reused in completely different ways if need be. Maybe initially, once those bits are done, I could write a little helper function the sits in pylinac that prompts the user to pip install the packages, and then provides a demo for use and setup in one's centre.

What do you think?

SimonBiggs commented:
Also, do you think it would be best to start a different module altogether within pylinac for the Elekta system? Or is it worth trying to make the logfile module vender agnostic?

jrkerns commented:
I fully encourage this project and at some point I'd be willing to roll it into pylinac. Unless the Elekta data structure is similar to Varian's, there's not much overlap from the existing code. I.e. all new code would have to be written to implement it, which is what you're doing right now. Given that I'm also swamped with clinical duties and pylinac has suffered because of it, you would also likely be frustrated if we needed to quickly adjust pylinac's code base. Please keep me abreast of this project. Do you have a few sample logs I could look at?

SimonBiggs commented:
@jrkerns, that sounds like a plan. I have a few examples of log files in the testing data directory of `trf2csv`:
https://gitlab.com/level-0/trf2csv/tree/master/trf2csv/tests/01_end_to_end_tests/data

samuelharmer commented:
Dear Simon,

We have five Elektas here in Bristol UK which I have access to. Let me know
if I can be of any help (testing, programming, etc). I'm a software
engineer by trade rather than physicist but I've been here long enough to
get the gist of RT ;-)

Feel free to email me off-list (me@samuelharmer.com).

Kind regards,
Samuel

On 8 November 2017 at 02:16, Simon Biggs <notifications@github.com> wrote:

> @jrkerns <https://github.com/jrkerns>, that sounds like a plan. I have a
> few examples of log files in the testing data directory of trf2csv:
> https://gitlab.com/level-0/trf2csv/tree/master/trf2csv/
> tests/01_end_to_end_tests/data
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jrkerns/pylinac/pull/107#issuecomment-342688198>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADlsMJ_929u6m6P0uI7FLcE2_kCtgZ1zks5s0Q7vgaJpZM4QUCB6>
> .
>

SimonBiggs commented:
Dear Samuel,

All of the above sounds amazing.

To get started would you be able to set up logfile extraction on at least one of your linacs? I outlined the steps we followed over at https://gitlab.com/level-0/trf2csv#setup-logfile-extraction. It will likely need to be done in coordination and with the approval of Elekta. If the Elekta representative wants someone to call I can give them the contact details of someone within Elekta Australia who went through the process with me. My email address is mail@simonbiggs.net.

If you're able to document and submit as a pull request any differences you underwent in setting up the logfile extraction that would be great.

As an aside, what planning system(s) and oncology information system do you use in Bristol?

Kind regards,
Simon

SimonBiggs commented:
@jrkerns as part of this project we are going to want to have some form of independent dose calculation algorithm. We are considering heading down the collapsed cone algorithm path.

Have you, or anyone you know, implemented or partially implemented a collapsed cone algorithm who would be willing to license the code under an AGPLv3+ license?

jrkerns commented:
The people I currently know (including myself) would likely not be able to do it under AGPL, but I will ask.

SimonBiggs commented:
For those following this, I need to change some of the wording in the trf2csv docs and possibly alter the recommended way to get the log files themselves.

At the moment I have removed the trf2csv off of GitLab, and I am sorry to say that I also removed it off of pypi. I am very sorry if that causes builds to break :(.

Please email me (me@simonbiggs.net) if you think you will find trf2csv useful, should I be allowed to once again release it online I'll make sure to let you know.

For your information, we are still using it internally. I have created a nice interface with https://github.com/SimonBiggs/scriptedforms which allows the selection of patient and field id for comparison between Mosaiq and Log files using an 'MU Fluence' similar to pylinac.

SimonBiggs commented:
@jrkerns I have implemented a version of "MU Fluence" which works for Elekta Agility:

https://github.com/CCA-Physics/pymedphys/blob/master/pymedphys/level1/mudensity.py

It works with logfiles, but also with control points exported from Mosaiq. It handles DMLC by determining the fraction of time a given pixel spends blocked by a leaf between control points.

It is still very much alpha, but hopefully there might be something that would plugin nicely to pylinac.

jrkerns commented:
@SimonBiggs This looks pretty cool and useful. I can't find any
documentation or examples though. Did I miss them somewhere?

On Mon, Jun 4, 2018 at 10:32 PM Simon Biggs <notifications@github.com>
wrote:

> @jrkerns <https://github.com/jrkerns> I have implemented a version of "MU
> Fluence" which works for Elekta Agility:
>
>
> https://github.com/CCA-Physics/pymedphys/blob/master/pymedphys/level1/mudensity.py
>
> It works with logfiles, but also with control points exported from Mosaiq.
> It handles DMLC by determining the fraction of time a given pixel spends
> blocked by a leaf between control points.
>
> It is still very much alpha, but hopefully there might be something that
> would plugin nicely to pylinac.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jrkerns/pylinac/pull/107#issuecomment-394561378>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AElLCIKsikQjrthk2EbTHEDfZxU1ofsJks5t5e2rgaJpZM4QUCB6>
> .
>

SimonBiggs commented:
It is very much missing those. It's also very much in alpha development
mode. It'll be a few more months at least before its API shows a hint of
settling down. I'll wait until that API is a bit more concrete before
investing the time into documenting the features. "Release early and often".

Hopefully by the time I present it at our Aussy conference later this year
it'll have some docs to boot.

We have an example of use in our private repo. I can invite you in to that
if you would like. You'll need a GitLab account though. And of course I am
happy to answer any questions you have if you did want to try taking it for
a spin.


On Wed., 6 Jun. 2018, 4:32 am James Kerns, <notifications@github.com> wrote:

> @SimonBiggs This looks pretty cool and useful. I can't find any
> documentation or examples though. Did I miss them somewhere?
>
> On Mon, Jun 4, 2018 at 10:32 PM Simon Biggs <notifications@github.com>
> wrote:
>
> > @jrkerns <https://github.com/jrkerns> I have implemented a version of
> "MU
> > Fluence" which works for Elekta Agility:
> >
> >
> >
> https://github.com/CCA-Physics/pymedphys/blob/master/pymedphys/level1/mudensity.py
> >
> > It works with logfiles, but also with control points exported from
> Mosaiq.
> > It handles DMLC by determining the fraction of time a given pixel spends
> > blocked by a leaf between control points.
> >
> > It is still very much alpha, but hopefully there might be something that
> > would plugin nicely to pylinac.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/jrkerns/pylinac/pull/107#issuecomment-394561378>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AElLCIKsikQjrthk2EbTHEDfZxU1ofsJks5t5e2rgaJpZM4QUCB6
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jrkerns/pylinac/pull/107#issuecomment-394814522>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGQVe5d3rNXzG6TiaO8D74Hjt0yUhMgXks5t5s7ZgaJpZM4QUCB6>
> .
>

jrkerns commented:
I'm on Gitlab and am thinking about moving completely over there. I'm a big fan of the company; I'm just working out a few kinks. For the time being, I have my github repos mirrored by Gitlab until I fully convert.

SimonBiggs commented:
Ahah, I see you https://gitlab.com/jrkerns/pylinac I'll add you to our
GitLab private repo. It's all still seriously early development.

On Wed, 6 Jun 2018 at 12:12 James Kerns <notifications@github.com> wrote:

> I'm on Gitlab and am thinking about moving completely over there. I'm a
> big fan of the company; I'm just working out a few kinks. For the time
> being, I have my github repos mirrored by Gitlab until I fully convert.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jrkerns/pylinac/pull/107#issuecomment-394917374>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGQVe2CMsC-oDEeXEFIpP-axJFliUhIwks5t5zp9gaJpZM4QUCB6>
> .
>

SimonBiggs commented:
For a rough get you started, the main function that is expected to be used
is the following:

https://github.com/CCA-Physics/pymedphys/blob/master/pymedphys/level1/mudensity.py#L348

mu, mlc, and jaw are all numpy arrays. mu is a one dimensional array, mlc a
three dimensional array, jaw is two dimensional.

The first dimension is always control point number. The last dimension is
mlc/jaw bank number. For mlc's the middle dimension is leaf pair number.

The coordinate system is taken to be bipolar to match the coordinate system
used by the Elekta linac service mode.

Distances are to be all in mm.

Hopefully that covers the basics.

If you would like I can supply an example set of arrays within the package.
That would probably be a really easy way on my end to help you get a feel
for what's going on.



On Wed., 6 Jun. 2018, 12:15 pm Simon Biggs, <me@simonbiggs.net> wrote:

> Ahah, I see you https://gitlab.com/jrkerns/pylinac I'll add you to our
> GitLab private repo. It's all still seriously early development.
>
> On Wed, 6 Jun 2018 at 12:12 James Kerns <notifications@github.com> wrote:
>
>> I'm on Gitlab and am thinking about moving completely over there. I'm a
>> big fan of the company; I'm just working out a few kinks. For the time
>> being, I have my github repos mirrored by Gitlab until I fully convert.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/jrkerns/pylinac/pull/107#issuecomment-394917374>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AGQVe2CMsC-oDEeXEFIpP-axJFliUhIwks5t5zp9gaJpZM4QUCB6>
>> .
>>
>

SimonBiggs commented:
@jrkerns alrighty, so I added a really basic example which includes the numpy arrays saved in a npz format, and within it is some commented out code that would plot the result:

https://github.com/CCA-Physics/pymedphys/blob/master/tests/mudensity/regression-test.py

Does that give enough of a lead to give it a try? (Can't really spend much more time on this just now, monthly linac QA to do :) ).

SimonBiggs commented:
Our next step in the coming weeks / month is to build up a testing suite for it. I wouldn't be surprised if there are some issues with the calculations. Hopefully we should be able to provide sufficient testing coverage with a range of test shape comparisons.
